optimize/cmaes.go

```diff
@@ -0,0 +1,381 @@
+		return NotTerminated, nil
+	case sd == 0:
+		sd = float64(cma.dim) * -36.8413614879 // ln(1e-16)
+	}
+	if cma.chol.LogDet() < sd {
+		return MethodConverge, nil
+	}
+	return NotTerminated, nil
+}
+
+func (cma *CmaEsChol) InitGlobal(dim, tasks int) int {
+	// Initialize the parameters
+	if cma.InitMean != nil && len(cma.InitMean) != dim {
+		panic("cma-es-chol: initial mean must be nil or have length equal to dimension")
+	}
+
+	// Set fixed algorithm parameters.
+	// Parameter values are from https://arxiv.org/pdf/1604.00772.pdf .
+	cma.dim = dim
+	cma.pop = cma.Population
+	if cma.pop == 0 {
+		cma.pop = 4 + int(3*math.Log(float64(dim))) // Note the implicit floor.
+	} else if cma.pop < 0 {
+		panic("cma-es-chol: negative population size")
+	}
+	mu := cma.pop / 2
+	cma.weights = resize(cma.weights, mu)
+	for i := range cma.weights {
+		v := math.Log(float64(mu)+0.5) - math.Log(float64(i)+1)
+		cma.weights[i] = v
+	}
+	floats.Scale(1/floats.Sum(cma.weights), cma.weights)
+	cma.muEff = 0
+	for _, v := range cma.weights {
+		cma.muEff += v * v
+	}
+	cma.muEff = 1 / cma.muEff
+
+	n := float64(dim)
+	cma.cc = (4 + cma.muEff/n) / (n + 4 + 2*cma.muEff/n)
+	cma.cs = (cma.muEff + 2) / (n + cma.muEff + 5)
+	cma.c1 = 2 / ((n+1.3)*(n+1.3) + cma.muEff)
+	cma.cmu = math.Min(1-cma.c1, 2*(cma.muEff-2+1/cma.muEff)/((n+2)*(n+2)+cma.muEff))
+	cma.ds = 1 + 2*math.Max(0, math.Sqrt((cma.muEff-1)/(n+1))-1) + cma.cs
+	// E[chi] is taken from https://en.wikipedia.org/wiki/CMA-ES (there
+	// listed as E[||N(0,1)||]).
+	cma.eChi = math.Sqrt(n) * (1 - 1.0/(4*n) + 1/(21*n*n))
+
+	// Allocate memory for function data.
+	cma.xs = mat.NewDense(cma.pop, dim, nil)
+	cma.fs = resize(cma.fs, cma.pop)
+
+	// Allocate and initialize adaptive parameters.
+	cma.sigma = cma.InitStepSize
+	if cma.sigma == 0 {
+		cma.sigma = 0.3
+	} else if cma.sigma < 0 {
+		panic("cma-es-chol: negative initial step size")
+	}
+	cma.pc = resize(cma.pc, dim)
+	for i := range cma.pc {
+		cma.pc[i] = 0
+	}
+	cma.ps = resize(cma.ps, dim)
+	for i := range cma.ps {
+		cma.ps[i] = 0
+	}
+	cma.mean = resize(cma.mean, dim)
+	if cma.InitMean != nil {
+		if len(cma.InitMean) != dim {
+			panic("cma-es-chol: InitMean length mismatch")
+		}
+		copy(cma.mean, cma.InitMean)
+	}
+	if cma.InitCholesky != nil {
+		cma.chol = mat.Cholesky{}
+		cma.chol.Clone(cma.InitCholesky)
+	} else {
+		// Set the initial Cholesky to I.
+		b := mat.NewDiagonal(dim, nil)
+		for i := 0; i < dim; i++ {
+			b.SetSymBand(i, i, 1)
+		}
+		var chol mat.Cholesky
+		ok := chol.Factorize(b)
+		if !ok {
+			panic("cma-es-chol: bad cholesky. shouldn't happen")
+		}
+		cma.chol = chol
+	}
+
+	cma.evals = make([]int, cma.pop)
+	for i := 0; i < cma.pop; i++ {
+		cma.evals[i] = i
+	}
+
+	cma.bestX = resize(cma.bestX, dim)
+	cma.bestF = math.Inf(1)
+
+	t := min(tasks, cma.pop)
+	cma.taskIdxs = make([]int, cma.pop)
+	for i := 0; i < t; i++ {
+		cma.taskIdxs[i] = -1
+	}
+	// Get a new waitgroup so that if the struct is re-used the.
```

Comment:
it seems we were interrupted mid-sentence :)

Reply:
Oh yea. Because I decided I should run the test before I spent the time writing the comment, and then the test failed :-) . Turned out to be a different bug.

Reply:
Done.